--- a/python/classy.pyx	2018-09-21 07:00:18.000000000 +0200
+++ b/python/classy.pyx	2019-06-06 16:54:44.517245217 +0200
@@ -971,8 +971,10 @@
     # This is commented because in the current form it only applies
     # to minimal LambdaCDM.
     # On would need to add contributions from ncdm, ddmdr, etc.
-    #def Omega_r(self):
-    #    return self.ba.Omega0_g+self.ba.Omega0_ur
+    # (JR) added again, in my understanding self.ba.Omega0_dcdmdr = self.ba.Omega0_dcdm + self.ba.Omega0_dr,
+    #    therefore only the latter contributes to Omega_r
+    def Omega_r(self):
+        return self.ba.Omega0_g+self.ba.Omega0_ur+self.ba.Omega0_dr
 
     def Omega_Lambda(self):
         return self.ba.Omega0_lambda
@@ -986,6 +988,21 @@
     def omega_b(self):
         return self.ba.Omega0_b * self.ba.h * self.ba.h
 
+    def Omega_dcdm(self):                   # (JR) added 
+        return self.ba.Omega0_dcdm
+    
+    def Omega_dr(self):                     # (JR) added 
+        return self.ba.Omega0_dr
+    
+    def Omega_dcdmdr(self):                     # (JR) added 
+        return self.ba.Omega0_dcdmdr
+
+    def Omega_ncdm_tot(self):                     # (JR) added 
+        return self.ba.Omega0_ncdm_tot
+    
+    def Omega_ur(self):                     # (JR) added 
+        return self.ba.Omega0_ur
+
     def Neff(self):
         return self.ba.Neff
 
@@ -1521,6 +1538,9 @@
             elif name == 'Omega_m':
                 value = (self.ba.Omega0_b + self.ba.Omega0_cdm+
                          self.ba.Omega0_ncdm_tot + self.ba.Omega0_dcdm)
+            elif name == 'Omega_m':
+                value = (self.ba.Omega0_b + self.ba.Omega0_cdm+
+                         self.ba.Omega0_ncdm_tot + self.ba.Omega0_dcdm)
             elif name == 'omega_m':
                 value = (self.ba.Omega0_b + self.ba.Omega0_cdm+
                          self.ba.Omega0_ncdm_tot + self.ba.Omega0_dcdm)/self.ba.h**2
--- a/python/cclassy_orig.pxd 2018-09-21 06:00:18.000000000 +0100
+++ b/python/cclassy.pxd  2019-07-03 12:50:04.697779053 +0100
@@ -61,6 +61,7 @@
         double cs2_fld
         double Omega0_ur
         double Omega0_dcdmdr
+        double Omega0_dr           # (JR) added
         double Omega0_scf
         double Omega0_k
         int bt_size
