--- a/src/bbn.c	2019-07-17 17:47:45.000000000 +0200
+++ b/src/bbn.c	2019-07-17 18:54:10.169706632 +0200
@@ -589,7 +589,7 @@
 		// SM and equivalent neutrinos share same temperature
 		rho_neuteq=2.*pow(pi,2.)/30.*7./8.*paramrelic->dNnu*pow(Tnu,4.);
 		P_neuteq=rho_neuteq/3.;
-		drho_neuteq=0.;
+		drho_neuteq=7.*pow(pi,2.)/30.*paramrelic->dNnu*pow(Tnu,3.);
 	}
 
 	/* ############### BARYON DENSITY ################ */
@@ -739,7 +739,11 @@
 	else paramerror->life_neutron=paramrelic->life_neutron;
 					
     int i;
-	memset(ratioH, 0.,sizeof(double) * (NNUC+1));
+	int ii;
+	for (ii=0;ii<NNUC+1;ii++)
+	{
+		ratioH[ii] = 0.;
+	}
 
 #ifdef OUTPUT
 	FILE *output;
@@ -1684,7 +1684,8 @@
 				if(paramrelic->phi_model&&rho_phi!=0.) rhophi_sav=rho_phi;
 			}
 			
-			for(int iloop=0;iloop<=11;iloop++)
+			int iloop;
+			for(iloop=0;iloop<=11;iloop++)
 			{
 				switch(iloop)
 				{
@@ -2345,8 +2349,15 @@
 {	
     int ie;    
     
-    memset(ratioH, 0.,sizeof(double) * (NNUC+1));
-    memset(cov_ratioH, 0.,sizeof(double) * (NNUC+1) * (NNUC+1));
+    int ii,  jj;
+    for(ii=0;ii<NNUC+1;ii++) 
+    {    
+        ratioH[ii] = 0.0;
+        for(jj=0;jj<NNUC+1;jj++)
+        {
+            cov_ratioH[ii][jj] = 0.0;
+        }
+    }    
 
     if(paramrelic->err==0)
     {
