--- a/source/thermodynamics.c
+++ b/source/thermodynamics.c
@@ -354,6 +354,15 @@
 	     "annihilation parameter suspiciously large (%e, while typical bounds are in the range of 1e-7 to 1e-6)",
 	     pth->annihilation);
 
+  if (pth->decay_fraction > 0) {
+    double lifetime = pba->tau_dcdm;
+    double bound = 4.6e-13 * pow((lifetime/1e13),-6.0) * pow((1.0+lifetime/1e13),7.0);
+    class_test(pth->decay_fraction > 1e3*bound,
+	       pth->error_message,
+	       "The value of 'pth->decay_fraction' (=%e) is suspiciously high given the choice of 'pba->tau_dcdm'. The bound on the decay fraction for 'pba->tau_dcdm'=%e is given by %e, which is exceeded by several orders of magnitude. We stop here since the code doesn't handle well with too high energy injection. Please adjust the inputs accordingly.",
+	       pth->decay_fraction, lifetime, bound);
+  }
+
   class_test(pth->PBH_evaporating_mass > 0 && pth->PBH_evaporating_mass < 1e15 && pth->PBH_fraction > 1e-4,pth->error_message,
 	     "The value of 'pth->PBH_fraction' that you enter is suspicious given the mass you chose. You are several orders of magnitude above the limit. The code doesn't handle well too high energy injection. Please choose  'pth->PBH_fraction < 1e-4'. ")
 
